src/optimal_image/search.rs

```diff
@@ -1,35 +1,127 @@
-use dataclients::ImageData;
+use dataclients::*;
+use image::*;
+use std::collections::HashMap;
+use std::error::Error;
 use std::path::Path;
+extern crate dssim;
+use EncodingConfig;
 use ImageConfig;
 use ImageFormat;
 use ImageScore;
 
+#[derive(Debug)]
 pub struct SearchOptions {
-    threshold: f64,
-    formats: Vec<ImageFormat>,
+    pub threshold: f64,
+    pub quality_range: (u8, u8),
+    pub formats: Vec<ImageFormat>,
 }
 
+pub type SearchResult = HashMap<ImageFormat, ImageConfig>;
+
 pub struct Search {
+    path: String,
     image_data: ImageData,
     options: SearchOptions,
+    result: Option<SearchResult>,
+    dssim: dssim::Dssim,
 }
 
 impl Search {
-    fn from_path(&self, path: &Path) -> &Search {
+    pub fn from_path(
+        path: &Path,
+        options: SearchOptions,
+    ) -> Result<Search, Box<Error>> {
         // TODO: initialize struct members based on path
-        self
+        let extension = path.extension().unwrap().to_str();
+
+        let original;
+        if extension == Some("jpg") {
+            let dataclient = Jpeg::new();
+            original = dataclient.load(path)?;
+        } else if extension == Some("png") {
+            let dataclient = Png::new();
+            original = dataclient.load(path)?;
+        } else {
+            return Err(Box::new(ImageError::FormatError(String::from(
+                "error",
+            ))));
+        }
+
+        let search = Search {
+            path: path.to_str().unwrap().to_string(),
+            image_data: original,
+            options: options,
+            result: None,
+            dssim: dssim::new(),
+        };
+
+        Ok(search)
+    }
+
+    pub fn run(&mut self) -> Option<SearchResult> {
+        // let mut config_map: HashMap<&ImageFormat, Vec<ImageConfig>> = HashMap::new();
+        let configs = self.options.formats.iter().fold(
+            HashMap::new(),
+            |mut map: HashMap<&ImageFormat, Vec<ImageConfig>>, format| {
+                let conf_list = (self.options.quality_range.0
```

Comment:
If I understood correctly, this is where we want to apply the logarithm to restrict the number of config combinations. Or do we want to make it incremental?

Reply:
the range includes all the steps from begin to end, e.g. 70 to 90. This loop will create all the configs for each step that could eventually be tried out. But in the binary search only the ones needed to get to the optimal result will be actually used. Again we can benchmark this once the whole things works, In theory this approach would be a little bit heavier on memory compared to lazily create the config, but would save some CPU per comparison run...